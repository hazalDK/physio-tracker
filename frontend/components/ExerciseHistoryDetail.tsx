import React, { useState } from "react";
import {
  View,
  Text,
  ScrollView,
  ActivityIndicator,
  TouchableOpacity,
  RefreshControl,
} from "react-native";
import tw from "tailwind-react-native-classnames";
import { useExerciseHistory } from "@/hooks/useExerciseHistory";
import { Ionicons } from "@expo/vector-icons";
import { useFocusEffect } from "@react-navigation/native";
import { useCallback } from "react";

// Component to show detailed exercise history
export default function ExerciseHistoryDetail() {
  const { loading, history, error, refreshHistory } = useExerciseHistory();
  const [expandedDate, setExpandedDate] = useState<string | null>(null);
  const [refreshing, setRefreshing] = useState(false);

  // Function to toggle expanded view for a date
  const toggleExpand = (date: string) => {
    if (expandedDate === date) {
      setExpandedDate(null);
    } else {
      setExpandedDate(date);
    }
  };

  // Refresh data when the screen comes into focus
  useFocusEffect(
    useCallback(() => {
      refreshHistory();
      return () => {
        // Cleanup if needed
      };
    }, [])
  );

  // Handle pull-to-refresh
  const onRefresh = useCallback(async () => {
    setRefreshing(true);
    await refreshHistory();
    setRefreshing(false);
  }, [refreshHistory]);

  // Loading state
  if (loading && !refreshing) {
    return (
      <View style={tw`flex-1 items-center justify-center bg-white`}>
        <ActivityIndicator size="large" color="#14b8a6" />
      </View>
    );
  }

  // Error state
  if (error) {
    return (
      <View style={tw`flex-1 items-center justify-center bg-white px-4`}>
        <Text style={tw`text-red-500 text-lg`}>{error}</Text>
        <TouchableOpacity
          style={tw`mt-4 bg-teal-500 py-2 px-4 rounded`}
          onPress={refreshHistory}
        >
          <Text style={tw`text-white`}>Try Again</Text>
        </TouchableOpacity>
      </View>
    );
  }

  // Empty state
  if (history.length === 0) {
    return (
      <ScrollView
        style={tw`flex-1 bg-white px-4 pt-2`}
        refreshControl={
          <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
        }
      >
        <View style={tw`flex-1 items-center justify-center h-full py-20`}>
          <Text style={tw`text-gray-500 text-lg text-center`}>
            No exercise history found. Complete some exercises to see your
            history here.
          </Text>
        </View>
      </ScrollView>
    );
  }

  return (
    <ScrollView
      style={tw`flex-1 bg-white px-4 pt-2`}
      refreshControl={
        <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
      }
    >
      <Text style={tw`text-xl mt-32 font-bold mb-4`}>Exercise History</Text>

      {history.map((item) => (
        <View
          key={item.date}
          style={tw`border-2 border-gray-200 rounded-lg p-4 mb-3`}
        >
          <TouchableOpacity
            style={tw`flex-row justify-between items-center`}
            onPress={() => toggleExpand(item.date)}
          >
            <View>
              <Text style={tw`font-bold text-lg`}>{item.formatted_date}</Text>
              <Text style={tw`text-gray-500`}>
                {item.exercises.length} exercise
                {item.exercises.length !== 1 ? "s" : ""} reported
              </Text>
              <Text style={tw`text-red-500`}>
                Overall Pain: {item.pain_level}/10
              </Text>
            </View>
            <Ionicons
              name={expandedDate === item.date ? "chevron-up" : "chevron-down"}
              size={24}
              color="#14b8a6"
            />
          </TouchableOpacity>

          {expandedDate === item.date && (
            <View style={tw`mt-3 pt-3 border-t border-gray-200`}>
              {item.exercises.map((exercise, exIndex) => (
                <View
                  key={`${item.date}-${exIndex}`}
                  style={tw`mb-3 pb-3 ${
                    exIndex < item.exercises.length - 1
                      ? "border-b border-gray-200"
                      : ""
                  }`}
                >
                  <Text style={tw`font-bold`}>{exercise.name}</Text>
                  <View style={tw`flex-row justify-between mt-1`}>
                    <Text>Sets: {exercise.sets}</Text>
                    <Text>Reps: {exercise.reps}</Text>
                    <Text
                      style={tw`${
                        exercise.pain > 3 ? "text-red-500" : "text-gray-600"
                      }`}
                    >
                      Pain: {exercise.pain}/10
                    </Text>
                  </View>
                </View>
              ))}
            </View>
          )}
        </View>
      ))}
    </ScrollView>
  );
}
