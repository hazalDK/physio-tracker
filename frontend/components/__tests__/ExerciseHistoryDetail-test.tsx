import * as React from "react";
import { render, fireEvent, waitFor, act } from "@testing-library/react-native";
import renderer from "react-test-renderer";

import ExerciseHistoryDetail from "../ExerciseHistoryDetail";

// Mock the custom hook
jest.mock("@/hooks/useExerciseHistory", () => ({
  useExerciseHistory: jest.fn(),
}));

// Mock the navigation hooks
jest.mock("@react-navigation/native", () => ({
  useFocusEffect: jest.fn((callback) => {
    // Execute the callback to simulate screen focus
    callback();
    return jest.fn();
  }),
}));

// Mock expo icons
jest.mock("@expo/vector-icons", () => ({
  Ionicons: "Ionicons-Mock",
}));

// Mock the tailwind
jest.mock("tailwind-react-native-classnames", () => ({
  __esModule: true,
  default: (styles: any) => styles,
}));

describe("ExerciseHistoryDetail Component", () => {
  // Setup the mock data for our tests
  const mockHistoryData = [
    {
      date: "2025-04-24",
      formatted_date: "April 24, 2025",
      pain_level: 3,
      exercises: [
        { name: "Squat", sets: 3, reps: 10, pain: 2 },
        { name: "Knee Extension", sets: 2, reps: 15, pain: 4 },
      ],
    },
    {
      date: "2025-04-23",
      formatted_date: "April 23, 2025",
      pain_level: 2,
      exercises: [{ name: "Leg raises", sets: 3, reps: 12, pain: 1 }],
    },
  ];

  // Mock implementation for useExerciseHistory hook
  beforeEach(() => {
    // Reset and setup our mock functions
    const { useExerciseHistory } = require("@/hooks/useExerciseHistory");

    // Default state - successful data load
    useExerciseHistory.mockReturnValue({
      loading: false,
      history: mockHistoryData,
      error: null,
      refreshHistory: jest.fn(),
    });
  });

  it("shows loading state", () => {
    const { useExerciseHistory } = require("@/hooks/useExerciseHistory");
    useExerciseHistory.mockReturnValue({
      loading: true,
      history: [],
      error: null,
      refreshHistory: jest.fn(),
    });

    const { getByTestId } = render(<ExerciseHistoryDetail />);
    expect(getByTestId("loading-indicator")).toBeDefined();
  });

  it("shows error state", () => {
    const { useExerciseHistory } = require("@/hooks/useExerciseHistory");
    useExerciseHistory.mockReturnValue({
      loading: false,
      history: [],
      error: "Failed to load exercise history",
      refreshHistory: jest.fn(),
    });

    const { getByText } = render(<ExerciseHistoryDetail />);
    expect(getByText("Failed to load exercise history")).toBeDefined();
    expect(getByText("Try Again")).toBeDefined();
  });

  it("shows empty state message when no history", () => {
    const { useExerciseHistory } = require("@/hooks/useExerciseHistory");
    useExerciseHistory.mockReturnValue({
      loading: false,
      history: [],
      error: null,
      refreshHistory: jest.fn(),
    });

    const { getByText } = render(<ExerciseHistoryDetail />);
    expect(
      getByText(
        "No exercise history found. Complete some exercises to see your history here."
      )
    ).toBeDefined();
  });

  it("expands and collapses exercise details when pressed", () => {
    const { getByText, queryByText } = render(<ExerciseHistoryDetail />);

    // Initially the detail should not be visible
    expect(queryByText("Squat")).toBeNull();

    // Press the date to expand
    fireEvent.press(getByText("April 24, 2025"));

    // Now the detail should be visible
    expect(getByText("Squat")).toBeDefined();
    expect(getByText("Sets: 3")).toBeDefined();

    // Press again to collapse
    fireEvent.press(getByText("April 24, 2025"));

    // Detail should be hidden again
    expect(queryByText("Squat")).toBeNull();
  });

  it("calls refreshHistory when pull-to-refresh is triggered", async () => {
    const mockRefreshHistory = jest.fn().mockResolvedValue(undefined);
    const { useExerciseHistory } = require("@/hooks/useExerciseHistory");
    useExerciseHistory.mockReturnValue({
      loading: false,
      history: mockHistoryData,
      error: null,
      refreshHistory: mockRefreshHistory,
    });

    const { getByTestId } = render(<ExerciseHistoryDetail />);
    const scrollView = getByTestId("scroll-view");

    // Trigger refresh
    fireEvent(scrollView, "refreshControl", { refreshing: false });

    await waitFor(() => {
      expect(mockRefreshHistory).toHaveBeenCalled();
    });
  });

  it("calls refreshHistory when Try Again button is pressed", () => {
    const mockRefreshHistory = jest.fn();
    const { useExerciseHistory } = require("@/hooks/useExerciseHistory");
    useExerciseHistory.mockReturnValue({
      loading: false,
      history: [],
      error: "Failed to load exercise history",
      refreshHistory: mockRefreshHistory,
    });

    const { getByText } = render(<ExerciseHistoryDetail />);
    const tryAgainButton = getByText("Try Again");

    fireEvent.press(tryAgainButton);

    expect(mockRefreshHistory).toHaveBeenCalled();
  });

  // Instead of testing styles directly, test the component rendering and behavior
  it("displays exercises with pain levels correctly", () => {
    const { getByText } = render(<ExerciseHistoryDetail />);

    // Expand the exercise list
    fireEvent.press(getByText("April 24, 2025"));

    // Check if both exercises are displayed
    expect(getByText("Squat")).toBeDefined();
    expect(getByText("Knee Extension")).toBeDefined();

    // Check if pain levels are displayed correctly
    expect(getByText("Pain: 2/10")).toBeDefined();
    expect(getByText("Pain: 4/10")).toBeDefined();

    // Verify our test data has the conditions we expect
    const lowPainExercise = mockHistoryData[0].exercises[0];
    const highPainExercise = mockHistoryData[0].exercises[1];

    expect(lowPainExercise.pain).toBeLessThanOrEqual(3);
    expect(highPainExercise.pain).toBeGreaterThan(3);
  });
});
