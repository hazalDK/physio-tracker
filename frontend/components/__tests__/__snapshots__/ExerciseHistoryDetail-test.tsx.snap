// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseHistoryDetail Component renders correctly with data 1`] = `null`;
